### ISSUE TITLE
padding and resampling is not commutative 

### ISSUE THREAD

**Issue by sheryjoe** [SCIInstitute/ShapeWorks#831]
**padding then resampling (incorrect result)**

```python
shapeSeg = sw.Image(filename)
shapeSegPadded = sw.Image(shapeSeg)
shapeSegPadded.pad(10, 0)
shapeSegResampled = sw.Image(shapeSegPadded)
shapeSegResampled.resample(sw.Vector(1,1,1), sw.InterpolationType.Linear )
```

```
Header information before padding: 
{
	dims: [52, 41, 21],
	origin: [-15, -11, -11],
	size: [52, 41, 42],
	spacing: [1, 1, 2]
}

Header information after padding: 
{
	dims: [52, 41, 21],
	origin: [-15, -11, -11],
	size: [52, 41, 42],
	spacing: [1, 1, 2]
}

Header information before resampling: 
{
	dims: [72, 61, 41],
	origin: [-15, -11, -11],
	size: [72, 61, 82],
	spacing: [1, 1, 2]
}

Header information after resampling: 
{
	dims: [72, 61, 82],
	origin: [-15, -11, -11],
	size: [72, 61, 82],
	spacing: [1, 1, 1]
}
```
visualization using pyvista with linked views
<img width="957" alt="image" src="https://user-images.githubusercontent.com/3253036/100491872-63349800-30e4-11eb-8049-6306fc664c1a.png">

In paraview, resampled (in red) has a shifted image boundaries  (resampled = red, padded = cyan)
<img width="1525" alt="image" src="https://user-images.githubusercontent.com/3253036/100492258-b65c1a00-30e7-11eb-9be5-d1edde181013.png">

**resampling then padding (correct result)**

```python
shapeSeg = sw.Image(filename)
shapeSegResampled = sw.Image(shapeSeg)
shapeSegResampled.resample(sw.Vector(1,1,1), sw.InterpolationType.Linear )

shapeSegPadded = sw.Image(shapeSegResampled)
shapeSegPadded.pad(10, 0)

```

```
Header information before resampling: 
{
	dims: [52, 41, 21],
	origin: [-15, -11, -11],
	size: [52, 41, 42],
	spacing: [1, 1, 2]
}

Header information after resampling: 
{
	dims: [52, 41, 42],
	origin: [-15, -11, -11],
	size: [52, 41, 42],
	spacing: [1, 1, 1]
}

Header information before padding: 
{
	dims: [52, 41, 42],
	origin: [-15, -11, -11],
	size: [52, 41, 42],
	spacing: [1, 1, 1]
}

Header information after padding: 
{
	dims: [72, 61, 62],
	origin: [-15, -11, -11],
	size: [72, 61, 62],
	spacing: [1, 1, 1]
}

```

pvista viz

<img width="959" alt="image" src="https://user-images.githubusercontent.com/3253036/100492093-6b8dd280-30e6-11eb-913f-bb6378b3c73c.png">

paraview (resampled = red, padded = cyan)

<img width="1525" alt="image" src="https://user-images.githubusercontent.com/3253036/100492121-bad40300-30e6-11eb-940b-523f0158e316.png">

**Comment by cchriste** [comment]
This doesn't use chaining and therefore isn't related to #861.

**Comment by cchriste** [comment]
The results are correct. 
Padding expands an images by the desired number _of voxels_, so resampling first results in smaller voxels and therefore less padding.

image before padding:
```
{
	dims: [97, 88, 42],
	origin: [-8, -6, -21],
	size: [97, 88, 84],
	spacing: [1, 1, 2]
}
```
```
img.pad(10, 0)  #pad by 10 voxels in every direction
{
	dims: [117, 108, 62],    # correct: each direction padded by 10 voxels
	origin: [-8, -6, -21],
	size: [117, 108, 124],   #correct: size of a voxel is 1x1x2, so new size is correct
	spacing: [1, 1, 2]
}
```
```
img.resample([1,1,1], sw.InterpolationType.Linear)
{
	dims: [117, 108, 124],    # correct original voxel 1x1x2, new voxel 1x1x1, so new dims are [x, y, 2*z]
	origin: [-8, -6, -21],
	size: [117, 108, 124],
	spacing: [1, 1, 1]
}
```

If volume is first resampled, the size of a voxel for padding is now smaller, so the final volume will be smaller.

**Comment by sheryjoe** [comment]
The origin of the image after padding then resample has changed.

**Comment by cchriste** [comment]
Fixed by dc000af2db62a9ba390061d374205d324496ed11, to be merged with #905 
<img width="849" alt="Screen Shot 2021-01-20 at 6 00 25 PM" src="https://user-images.githubusercontent.com/1260945/105260399-fde0ae80-5b4a-11eb-9b3b-bab9924fc3b2.png">

**Comment by archanasri** [comment]
Fix merged to master.
